File: android/sdui-core/src/main/java/com/nba/sdui/core/renderer/atomic/AtomicRouter.kt
package com.nba.sdui.core.renderer.atomic

import android.util.Log
import androidx.compose.runtime.Composable
import com.nba.sdui.core.models.generated.AtomicElement
import com.nba.sdui.core.state.SduiAction

private const val MAX_TREE_DEPTH = 6

/**
 * AtomicRouter — dispatches an [AtomicElement] to the correct atomic
 * primitive composable. The router itself is styling-free: every
 * box-model concern (margin, padding, background, cornerRadius,
 * shadow, border, opacity, width / height / fillWidth, variant chrome,
 * badge overlay) is applied by [AtomicBox] inside each primitive.
 *
 * This is the atomic counterpart to SectionRouter. A defensive [depth]
 * guard prevents malformed payloads from causing stack overflows or
 * deep Compose measure-pass issues. Server-side validation is the
 * primary enforcement; this is a safety net for stale caches or
 * manual JSON authoring.
 *
 * [onStateChange] is threaded through so that SectionSlot elements can
 * pass it to SectionRouter for stateful sections (Form, TabGroup).
 *
 * [sectionSlotDepth] tracks how many SectionSlot → AtomicComposite
 * cycles have occurred. See [AtomicSectionSlot] for the recursion
 * guard.
 */
@Composable
fun AtomicRouter(
    element: AtomicElement,
    screenState: Map<String, Any>,
    onAction: (SduiAction) -> Unit,
    depth: Int = 0,
    onStateChange: (String, Any) -> Unit = { _, _ -> },
    sectionSlotDepth: Int = 0
) {
    if (depth > MAX_TREE_DEPTH) {
        Log.w("AtomicRouter", "Max tree depth ($MAX_TREE_DEPTH) exceeded — skipping element: ${element.type}")
        return
    }
    when (element.type) {
        "Container"       -> AtomicContainer(element, screenState, onAction, depth, onStateChange, sectionSlotDepth)
        "Text"            -> AtomicText(element, screenState, onAction)
        "Image"           -> AtomicImage(element, screenState, onAction)
        "Button"          -> AtomicButton(element, screenState, onAction)
        "Spacer"          -> AtomicSpacer(element)
        "Divider"         -> AtomicDivider(element, screenState, onAction)
        "ScrollContainer" -> AtomicScrollContainer(element, screenState, onAction, depth, onStateChange, sectionSlotDepth)
        "Conditional"     -> AtomicConditional(element, screenState, onAction, depth, onStateChange, sectionSlotDepth)
        "DisplayGrid"     -> AtomicDisplayGrid(element, screenState, onAction)
        "SectionSlot"     -> AtomicSectionSlot(element, screenState, onAction, onStateChange, sectionSlotDepth)
        else              -> Log.w("AtomicRouter", "Unknown atomic element type: ${element.type}")
    }
}

File: ios/Sources/SduiCore/Rendering/AtomicRouter.swift
import SwiftUI
import os

private let logger = Logger(subsystem: "com.nba.sdui", category: "AtomicRouter")

private let maxDepth = 6

/// Recursive dispatcher for atomic UI elements. Enforces the atomic-tree
/// performance contract: max depth 6, max children 20, max nodes 50.
/// Unknown element types are silently skipped with a warning log — same
/// shape as the Android and web atomic routers.
///
/// Box-model concerns (margin, padding, background, cornerRadius, shadow,
/// border, opacity, width, height, fillWidth, badge, variant chrome) are
/// applied uniformly by `AtomicBoxModifier` inside each primitive
/// renderer, so this router is dispatch-only.
struct AtomicRouter: View {
    let element: AtomicElement
    let screenState: ScreenState
    let onAction: (Action) -> Void
    let depth: Int

    var body: some View {
        if depth > maxDepth {
            let _ = logger.warning("Max tree depth exceeded at depth \(depth) — skipping element")
            EmptyView()
        } else {
            routeElement()
        }
    }

    @ViewBuilder
    private func routeElement() -> some View {
        switch element.type {
        case "Container":
            AtomicContainerView(element: element, screenState: screenState, onAction: onAction, depth: depth + 1)

        case "Text":
            AtomicTextView(element: element, screenState: screenState, onAction: onAction)

        case "Image":
            AtomicImageView(element: element, screenState: screenState, onAction: onAction)

        case "Button":
            AtomicButtonView(element: element, screenState: screenState, onAction: onAction)

        case "Spacer":
            AtomicSpacerView(element: element)

        case "Divider":
            AtomicDividerView(element: element)

        case "ScrollContainer":
            AtomicScrollContainerView(element: element, screenState: screenState, onAction: onAction, depth: depth + 1)

        case "Conditional":
            AtomicConditionalView(element: element, screenState: screenState, onAction: onAction, depth: depth + 1)

        case "DisplayGrid":
            AtomicDisplayGridView(element: element)

        case "SectionSlot":
            AtomicSectionSlotView(element: element, screenState: screenState, onAction: onAction)

        case "LiveClock":
            AtomicLiveClockView(element: element)

        default:
            EmptyView()
                .onAppear {
                    logger.warning("skipping unknown atomic element type=\(element.type, privacy: .public) id=\(element.id ?? "(nil)", privacy: .public)")
                }
        }
    }
}
